In [1]:
import numpy as np
from scipy import stats

### 1. Провести дисперсионный анализ для определения того, 
#### есть ли различия среднего роста среди
#### взрослых футболистов, хоккеистов и штангистов.
#### Даны значения роста в трех группах случайно выбранных спортсменов:
    
    Футболисты: 173, 175, 180, 178, 177, 185, 183, 182.
    
    Хоккеисты: 177, 179, 180, 188, 177, 172, 171, 184, 180.
    
    Штангисты: 172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170.
    
#### $H_0$ - различия среднего роста среди взрослых футболистов, хоккеистов и штангистов НЕТ 
#### то есть $(H_0: \overline{footballers} = \overline{hockey\_players} =  \overline{weightlifters})$

#### Выборка роста футболистов

In [2]:
footballers =  np.array([173, 175, 180, 178, 177, 185, 183, 182], dtype=np.float64)
n_f = len(footballers)
footballers_mean = np.mean(footballers)


Величина выборки, среднее значение по выборке

In [3]:
n_f, footballers_mean

(8, 179.125)

#### Выборка роста хоккеистов

In [4]:
hockey_players =  np.array([177, 179, 180, 188, 177, 172, 171, 184, 180], dtype=np.float64)
n_h = len(hockey_players)
hockey_players_mean = np.mean(hockey_players)


Величина выборки, среднее значение по выборке

In [5]:
n_h, hockey_players_mean

(9, 178.66666666666666)

#### Выборка роста штангистов

In [6]:
weightlifters = np.array([172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170], dtype=np.float64)
n_w = len(weightlifters)
weightlifters_mean = np.mean(weightlifters)


Величина выборки, среднее значение по выборке

In [7]:
n_w, weightlifters_mean

(11, 172.72727272727272)

#### Общая выборка роста спортсменнов трех вышестоящих выборок

In [8]:
all_athletes = np.concatenate([footballers, hockey_players, weightlifters])
k_all = len([footballers, hockey_players, weightlifters])
n_all = len(all_athletes)
all_athletes_mean = np.mean(all_athletes)


Количество групп выборок, величина всей выборки спортсменов, среднее значение по всей выборке спортсменос

In [9]:
k_all, n_all, all_athletes_mean

(3, 28, 176.46428571428572)

#### Сумма квадратов отклонений:

In [10]:
sum_of_squares_of_deviations_of_observations = np.sum((all_athletes - all_athletes_mean)**2)
sum_of_squares_of_deviations_of_observations

830.9642857142854

#### Сумма квадратов отклонений средних групповых значений от общего среднего:

In [11]:
s2_f = ((footballers_mean - all_athletes_mean)**2) * n_f + ((hockey_players_mean - all_athletes_mean)**2) * n_h + ((weightlifters_mean - all_athletes_mean)**2) * n_w
s2_f

253.9074675324678

#### Остаточная сумму квадратов отклонений

In [12]:
s2_residual = np.sum((footballers - footballers_mean)**2 ) + np.sum((hockey_players - hockey_players_mean)**2) + np.sum((weightlifters - weightlifters_mean)**2)       
s2_residual  

577.0568181818182

#### Общая дисперсия:

In [13]:
sigma2_general = sum_of_squares_of_deviations_of_observations / (n_all - 1)
sigma2_general

30.776455026455015

#### Факторная дисперсия:

In [14]:
sigma2_f = s2_f / (k_all - 1)
sigma2_f

126.9537337662339

#### Остаточная дисперсия:

In [15]:
sigma2_residual = s2_residual / (n_all - k_all)
sigma2_residual

23.08227272727273

$F_H =\frac{\sigma_{F}^{2}}{\sigma_{residual}^{2}}$

In [16]:
F_H = sigma2_f/sigma2_residual
F_H

5.500053450812598

Факторная степеней свободы $df_{f}=k-1$, остаточная степеней свободы $df_{residual}=n-k$

In [17]:
df_f = k_all- 1
df_residual = n_all - k_all
df_f, df_residual

(2, 25)

Найдем значение $F_{k}$ 
в таблице критических точек распределения Фишера-Снедекора для заданного уровня значимости $\alpha=0.05$ и $\nu_1 = df_{f} = 2, \quad \nu_2 = df_{residual}= 25$:

#### Для данных значений $F_{k} = 3.39$. Так как $F_H > F_{k}$, различие среднего роста в трех группах статистически значимо.

### $H_0$ - Отвергаем

Эмпирическое корреляционное отношение $\eta^2$:

In [18]:
eta2 = s2_f / sum_of_squares_of_deviations_of_observations
eta2

0.30555761769498

Так как принято считать, что при значениях $\eta^2$ ниже 0.2-0.3 групповые значения
средних не имеют статистически достоверного отличия.

### Вычесленое значение $\eta^2 = 0.30555761769498$ то можно считать что групповые значения средних имеют статистически достоверные отличия $H_0$ - Отвергаем